src/dialogs/PreviewImageOverlay.cc

```diff
@@ -0,0 +1,121 @@
+ */
+
+#include <QApplication>
+#include <QBuffer>
+#include <QDebug>
+#include <QFile>
+#include <QHBoxLayout>
+#include <QVBoxLayout>
+
+#include "Config.h"
+
+#include "dialogs/PreviewImageOverlay.h"
+
+using namespace dialogs;
+
+PreviewImageOverlay::PreviewImageOverlay(QWidget *parent)
+  : QWidget{parent}
+  , titleLabel_{tr("Upload image?"), this}
+  , imageLabel_{this}
+  , imageName_{tr("clipboard"), this}
+  , upload_{tr("Upload"), this}
+  , cancel_{tr("Cancel"), this}
+{
+        auto hlayout = new QHBoxLayout;
+        hlayout->addWidget(&upload_);
+        hlayout->addWidget(&cancel_);
+
+        auto vlayout = new QVBoxLayout{this};
+        vlayout->addWidget(&titleLabel_);
+        vlayout->addWidget(&imageLabel_);
+        vlayout->addWidget(&imageName_);
+        vlayout->addLayout(hlayout);
+
+        connect(&upload_, &QPushButton::clicked, [&]() {
+                emit confirmImageUpload(imageData_, imageName_.text());
+                close();
+        });
+        connect(&cancel_, &QPushButton::clicked, [&]() { close(); });
+}
+
+void
+PreviewImageOverlay::init()
+{
+        auto window   = QApplication::activeWindow();
+        auto winsize  = window->frameGeometry().size();
+        auto center   = window->frameGeometry().center();
+        auto img_size = image_.size();
+
+        imageName_.setText(imagePath_);
+
+        setAutoFillBackground(true);
+        setWindowFlags(Qt::Tool | Qt::WindowStaysOnTopHint);
+        setWindowModality(Qt::WindowModal);
+
+        titleLabel_.setStyleSheet(
+          QString{"font-weight: bold; font-size: %1px;"}.arg(conf::headerFontSize));
+        titleLabel_.setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+        titleLabel_.setAlignment(Qt::AlignCenter);
+        imageLabel_.setSizePolicy(QSizePolicy::Maximum, QSizePolicy::Maximum);
+        imageLabel_.setAlignment(Qt::AlignCenter);
+        imageName_.setSizePolicy(QSizePolicy::Minimum, QSizePolicy::Minimum);
+        imageName_.setAlignment(Qt::AlignCenter);
+        upload_.setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
+        cancel_.setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
+        upload_.setFontSize(conf::btn::fontSize);
+        cancel_.setFontSize(conf::btn::fontSize);
+
+        // Scale image preview to the size of the current window if it is larger.
+        if ((img_size.height() * img_size.width()) > (winsize.height() * winsize.width())) {
+                imageLabel_.setPixmap(image_.scaled(winsize, Qt::KeepAspectRatio));
+        } else {
+                imageLabel_.setPixmap(image_);
+                move(center.x() - (width() * 0.5), center.y() - (height() * 0.5));
+        }
+        imageLabel_.setScaledContents(false);
+
+        raise();
+}
+
+void
+PreviewImageOverlay::setImageAndCreate(const QByteArray &data, const QString &type)
+{
+        imageData_ = data;
+        image_.loadFromData(imageData_);
```

Comment:
It would be nice to have a fallback (perhaps a QLabel with some explanatory text) for when this fails, i.e. when the user pasted in an image of a type that Qt doesn't support decoding, or that `QImage` can't accurately guess the format of. In those cases we should still allow uploading, but the current behavior might produce a confusing UI.